File: src/lib/edge-workers/confirmation.ts
import UserError from '../user-error';
import { escapeTerminalText } from './output';

export interface ProductionMutationConfirmationRequest {
	action: 'deploy' | 'enable';
	appName: string;
	envType: string;
	workerNames: readonly string[];
	enableAfterDeploy: boolean;
	skipConfirmation: boolean;
	nonInteractive: boolean;
}

export interface EdgeWorkerDeletionConfirmationRequest {
	appName: string;
	envType: string;
	workerName: string;
	force: boolean;
}

export type EdgeWorkerConfirmFunction = ( message: string ) => Promise< boolean >;

export function isInteractiveEdgeWorkers( options: { nonInteractive?: boolean } ): boolean {
	return (
		process.env.VIP_NON_INTERACTIVE !== '1' &&
		! options.nonInteractive &&
		Boolean( process.stdout.isTTY )
	);
}

function productionMutationConfirmationMessage(
	request: ProductionMutationConfirmationRequest
): string {
	const target = `${ escapeTerminalText( request.appName ) }.${ escapeTerminalText(
		request.envType
	) }`;
	if ( request.action === 'enable' ) {
		return `Enable edge worker "${ escapeTerminalText(
			request.workerNames[ 0 ]
		) }" on ${ target }?`;
	}

	const action = request.enableAfterDeploy ? 'Deploy and enable' : 'Deploy';
	const workerLabel = request.workerNames.length === 1 ? 'edge worker' : 'edge workers';
	const preposition = request.enableAfterDeploy ? 'on' : 'to';
	const workerNames = request.workerNames.map( escapeTerminalText ).join( ', ' );

	return `${ action } ${ request.workerNames.length } ${ workerLabel } (${ workerNames }) ${ preposition } ${ target }?`;
}

export async function confirmProductionEdgeWorkerMutation(
	request: ProductionMutationConfirmationRequest,
	confirmFn: EdgeWorkerConfirmFunction
): Promise< void > {
	if ( request.envType !== 'production' || request.skipConfirmation ) {
		return;
	}

	if ( request.nonInteractive ) {
		const action =
			request.action === 'deploy' && request.enableAfterDeploy
				? 'deploy and enable'
				: request.action;
		throw new UserError(
			`Refusing to ${ action } edge workers in production without confirmation. ` +
				'Pass --skip-confirmation to proceed non-interactively.'
		);
	}

	const message = productionMutationConfirmationMessage( request );

	if ( ! ( await confirmFn( message ) ) ) {
		throw new UserError( 'Command cancelled by user.' );
	}
}

export async function confirmEdgeWorkerDeletion(
	request: EdgeWorkerDeletionConfirmationRequest,
	confirmFn: EdgeWorkerConfirmFunction
): Promise< void > {
	if ( request.force ) {
		return;
	}

	const confirmed = await confirmFn(
		`Permanently delete edge worker "${ escapeTerminalText(
			request.workerName
		) }" from ${ escapeTerminalText( request.appName ) }.${ escapeTerminalText(
			request.envType
		) }?`
	);
	if ( ! confirmed ) {
		throw new UserError( 'Command cancelled by user.' );
	}
}

File: internal/edgeworkers/confirmation.go
package edgeworkers

import (
	"errors"
	"fmt"
	"strings"
)

type ProductionConfirmation struct {
	Action, AppName, EnvType                            string
	WorkerNames                                         []string
	EnableAfterDeploy, SkipConfirmation, NonInteractive bool
}

func ConfirmProduction(req ProductionConfirmation, confirm func(string) (bool, error)) error {
	if req.EnvType != "production" || req.SkipConfirmation {
		return nil
	}
	action := req.Action
	if action == "deploy" && req.EnableAfterDeploy {
		action = "deploy and enable"
	}
	if req.NonInteractive {
		return fmt.Errorf("Refusing to %s edge workers in production without confirmation. Pass --skip-confirmation to proceed non-interactively.", action)
	}
	target := EscapeTerminalText(req.AppName) + "." + EscapeTerminalText(req.EnvType)
	var message string
	if req.Action == "enable" {
		if len(req.WorkerNames) == 0 {
			return errors.New("No worker selected.")
		}
		message = fmt.Sprintf("Enable edge worker \"%s\" on %s?", EscapeTerminalText(req.WorkerNames[0]), target)
	} else {
		verb, preposition, label := "Deploy", "to", "edge workers"
		if req.EnableAfterDeploy {
			verb = "Deploy and enable"
			preposition = "on"
		}
		if len(req.WorkerNames) == 1 {
			label = "edge worker"
		}
		message = fmt.Sprintf("%s %d %s (%s) %s %s?", verb, len(req.WorkerNames), label, escapedJoin(req.WorkerNames, ", ", ""), preposition, target)
	}
	return confirmOrCancel(message, confirm)
}
func confirmOrCancel(message string, confirm func(string) (bool, error)) error {
	ok, err := confirm(message)
	if err != nil {
		return err
	}
	if !ok {
		return errors.New("Command cancelled by user.")
	}
	return nil
}
func ConfirmDeletion(app, env, name string, force bool, confirm func(string) (bool, error)) error {
	if force {
		return nil
	}
	return confirmOrCancel(fmt.Sprintf("Permanently delete edge worker \"%s\" from %s.%s?", EscapeTerminalText(name), EscapeTerminalText(app), EscapeTerminalText(env)), confirm)
}
func escapedJoin(values []string, separator, fallback string) string {
	parts := make([]string, len(values))
	for i, s := range values {
		parts[i] = EscapeTerminalText(s)
	}
	joined := strings.Join(parts, separator)
	if joined == "" {
		return fallback
	}
	return joined
}
